### PR TITLE
prov/efa: Remove fi_sendmsg/recvmsg but still use efa_ep's qp and av

### PR DESCRIPTION
**Pull request by sunkuamzn**
For SHM EP, use fi_sendv and fi_recvv
For EFA EP
* Allocate ibv_send_wr and ibv_recv_wr in rxr_pkt_entry
* Still use efa_ep->qp, efa_ep->av and efa_ep->xmit_more_wr*
* Directly call efa_post_flush or ibv_post_recv in rxr_pkt_entry.c

Signed-off-by: Sai Sunku <sunkusa@amazon.com>

### CODE REVIEW

File: prov/efa/src/efa.h [hunk 1]
```diff
@@ -231,16 +231,6 @@ struct efa_ep {
 	bool			util_ep_initialized;
 };
 
-struct efa_send_wr {
```
Comment (wzamazon) [inline prov/efa/src/efa.h:234]: Why move the definitinion? it is still used by dgram endpoint, and should remain here.

File: prov/efa/src/rxr/rxr_cq.c [hunk 2]
```diff
@@ -310,6 +310,9 @@ void rxr_cq_queue_rnr_pkt(struct rxr_ep *ep,
 #endif
 	dlist_insert_tail(&pkt_entry->entry, list);
 
+	// Remove other packets that were sent with the packet that encountered RNR
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_cq.c:313]: Should be `/* ... */`

The comment is a little confusing. maybe explain what `pkt_entry->send_wr.wr.next` is

File: prov/efa/src/efa_msg.c [hunk 3]
```diff
@@ -286,7 +286,6 @@ static void efa_post_send_sgl(struct efa_ep *ep, const struct fi_msg *msg,
 	struct efa_mr *efa_mr;
 	struct ibv_send_wr *wr = &ewr->wr;
 	struct ibv_sge *sge;
-	size_t sgl_idx = 0;
```
Comment (wzamazon) [inline prov/efa/src/efa_msg.c:289]: This change is good. But probably should go into a separate commit

File: prov/efa/src/efa_msg.c [hunk 4]
```diff
@@ -391,7 +392,7 @@ static ssize_t efa_post_send(struct efa_ep *ep, const struct fi_msg *msg, uint64
 	if (flags & FI_MORE)
 		return 0;
 
-	ret = efa_post_flush(ep, &bad_wr);
+	ret = efa_post_flush(ep, &bad_wr, true);
```
Comment (wzamazon) [inline prov/efa/src/efa_msg.c:395]: add `true /* free */` to indicate is meaning

File: prov/efa/src/rxr/rxr_pkt_entry.h [hunk 5]
```diff
@@ -114,6 +114,17 @@ struct rxr_pkt_sendv {
  * is not registered (when it is unexpected or out-of-order). A new packet entry will be cloned
  * using endpoint's read_copy_pkt_pool, whose memory was registered.
  */
+
+struct efa_send_wr {
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_entry.h:118]: As mentioned before, this should remain in `efa.h` because it used by both rdm and dgram

**Review by wzamazon — CHANGES_REQUESTED** [review]
With the change to `efa_send_wr` and `efa_recv_wr` you need to update  how `efa_ep->send_wr_bufpool` and `efa_ep->recv_wr_pool` is created. Specifically the entry size should be just `sizeof(efa_send_wr)` and `sizeof(efa_recv_wr)`

File: prov/efa/src/rxr/rxr_pkt_entry.h (hunk 5 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_pkt_entry.h:118]: The import hierarchy right now is that `efa.h` imports `rxr.h` and `rxr.h` does not import `efa.h`. So if we keep this in `efa.h`, we have to add `#import "efa.h"` to `rxr_pkt_entry.h` which leads to circular dependency issues and compilation failure

I can fix the import hierarchy in a later commit when we rename `efa` to `efa_dgram`, `rxr` to `efa_rdm` and move common code to common files.

**Comment by sunkuamzn** [comment]
> With the change to `efa_send_wr` and `efa_recv_wr` you need to update how `efa_ep->send_wr_bufpool` and `efa_ep->recv_wr_pool` is created. Specifically the entry size should be just `sizeof(efa_send_wr)` and `sizeof(efa_recv_wr)`

But the DGRAM provider needs memory for the SGEs if a higher applications calls the DGRAM provider directly

File: prov/efa/src/efa_msg.c (hunk 3 above)
Comment (sunkuamzn) [reply prov/efa/src/efa_msg.c]: ACK

File: prov/efa/src/efa_msg.c (hunk 4 above)
Comment (sunkuamzn) [reply prov/efa/src/efa_msg.c]: done

File: prov/efa/src/rxr/rxr_pkt_entry.h (hunk 5 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_entry.h:118]: In that case, please add a todo comment here.

File: prov/efa/src/rxr/rxr_pkt_entry.h [hunk 6]
```diff
@@ -114,6 +114,17 @@ struct rxr_pkt_sendv {
...
+	struct ibv_send_wr wr;
+	struct ibv_sge sge[2]; // EFA device supports a maximum of 2 iov/SGE
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_entry.h:120]: "//" should be "/**< ... */" for doxygen comment

Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_pkt_entry.h]: done

File: prov/efa/src/rxr/rxr_pkt_entry.h (hunk 5 above)
Comment (sunkuamzn) [reply prov/efa/src/rxr/rxr_pkt_entry.h]: done

**Review by wzamazon — APPROVED** [review]
